packages/override-tools/src/GitReactFileRepository.ts

```diff
@@ -51,26 +57,73 @@ export default class GitReactFileRepository
     filename: string,
     reactNativeVersion: string,
   ): Promise<string | null> {
-    // We have a potential race condition where one call to getFileContents
-    // could checkout out a new tag while an existing call is rading a file.
-    // Queue items to ensure the read operation is performed atomically
-    return this.actionQueue.enqueue(() => {
-      return this.getFileContentsNonAtomic(filename, reactNativeVersion);
+    return this.actionQueue.enqueue(async () => {
+      await this.checkoutVersion(reactNativeVersion);
+      const filePath = path.join(this.gitDirectory, filename);
+
+      try {
+        return (await fs.promises.readFile(filePath)).toString();
+      } catch {
+        return null;
+      }
     });
   }
 
-  private async getFileContentsNonAtomic(
+  async generatePatch(
     filename: string,
     reactNativeVersion: string,
-  ): Promise<string | null> {
-    await this.checkoutVersion(reactNativeVersion);
-    const filePath = path.join(this.gitDirectory, filename);
+    newContent: string,
+  ): Promise<string> {
+    return this.actionQueue.enqueue(async () => {
+      await this.checkoutVersion(reactNativeVersion);
+      const filePath = path.join(this.gitDirectory, filename);
 
-    try {
-      return (await fs.promises.readFile(filePath)).toString();
-    } catch {
-      return null;
-    }
+      try {
+        await fs.promises.writeFile(filePath, newContent);
+        const patch = await this.gitClient.diff([
+          '--patch',
+          '--ignore-space-at-eol',
+        ]);
+        return patch;
+      } finally {
+        await this.gitClient.reset('hard');
+      }
+    });
+  }
+
+  async applyPatchToFile(
```

Comment:
nit: I should document this and the above. The intentional behavior that this doesn't permanently modify local state could be a bit confusing.